src/coprocessor/codec/convert.rs

```diff
@@ -43,53 +100,161 @@ pub fn truncate_f64(mut f: f64, flen: u8, decimal: u8) -> Res<f64> {
+            ERR_DATA_OUT_OF_RANGE,
+        )
     }};
 }
 
-/// `convert_uint_to_int` converts an uint value to an int value.
+/// Converts an int value to a different int value.
 #[inline]
-pub fn convert_uint_to_int(val: u64, upper_bound: i64, tp: FieldTypeTp) -> Result<i64> {
+pub fn convert_int_to_int(ctx: &mut EvalContext, val: i64, tp: FieldTypeTp) -> Result<i64> {
+    let lower_bound = integer_signed_lower_bound(tp);
+    // https://dev.mysql.com/doc/refman/8.0/en/out-of-range-and-overflow.html
+    if val < lower_bound {
+        ctx.handle_overflow(overflow!(val, lower_bound))?;
+        return Ok(lower_bound);
+    }
+    let upper_bound = integer_signed_upper_bound(tp);
+    if val > upper_bound {
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
+    }
+    Ok(val)
+}
+
+/// Converts an uint value to an int value.
+#[inline]
+pub fn convert_uint_to_int(ctx: &mut EvalContext, val: u64, tp: FieldTypeTp) -> Result<i64> {
+    let upper_bound = integer_signed_upper_bound(tp);
     if val > upper_bound as u64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
     }
     Ok(val as i64)
 }
 
-/// `convert_float_to_int` converts an f64 value to an i64 value.
-///  Returns the overflow error if the value exceeds the boundary.
-pub fn convert_float_to_int(
-    fval: f64,
-    lower_bound: i64,
-    upper_bound: i64,
-    tp: FieldTypeTp,
-) -> Result<i64> {
+/// Converts an f64 value to an i64 value.
+#[inline]
+pub fn convert_float_to_int(ctx: &mut EvalContext, fval: f64, tp: FieldTypeTp) -> Result<i64> {
     // TODO any performance problem to use round directly?
     let val = fval.round();
+    let lower_bound = integer_signed_lower_bound(tp);
     if val < lower_bound as f64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, lower_bound))?;
+        return Ok(lower_bound);
     }
 
+    let upper_bound = integer_signed_upper_bound(tp);
     if val > upper_bound as f64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
     }
     Ok(val as i64)
 }
 
-/// `convert_float_to_uint` converts a f64 value to a u64 value.
-/// Returns the overflow error if the value exceeds the boundary.
-pub fn convert_float_to_uint(fval: f64, upper_bound: u64, tp: FieldTypeTp) -> Result<u64> {
+/// Converts a byte arrays to an i64 in best effort.
+pub fn convert_bytes_to_int(ctx: &mut EvalContext, bytes: &[u8], tp: FieldTypeTp) -> Result<i64> {
+    let s = match str::from_utf8(bytes) {
+        Ok(s) => s,
+        Err(err) => {
+            ctx.handle_truncate(true)?;
+            let (valid, _) = bytes.split_at(err.valid_up_to());
+            unsafe { str::from_utf8_unchecked(valid) }
+        }
+    };
+    let s = s.trim();
+    let vs = get_valid_int_prefix(ctx, s)?;
```

Comment:
Looks like the logic should be:

1. Use `String::from_utf8_lossy` to get a string.
2. Get valid int prefix ...

Reply:
There is no benefit to using `String::from_utf8_lossy`. New memory allocation maybe appears if the string contains invalid utf8 characters.
See: https://doc.rust-lang.org/src/alloc/string.rs.html#571